src/hotspot/cpu/riscv/assembler_riscv.hpp

```diff
@@ -1959,6 +2007,970 @@ enum Nf {
 
 #undef INSN
 
+// ========================================
+// RISC-V Compressed Instructions Extension
+// ========================================
+// Note:
+//   1. When UseRVC is enabled, 32-bit instructions under 'CompressibleRegion's will be
+//      transformed to 16-bit instructions if compressible.
+//   2. RVC instructions in Assembler always begin with 'c_' prefix, as 'c_li',
+//      but most of time we have no need to explicitly use these instructions.
+//   3. We introduce 'CompressibleRegion' to hint instructions in this Region's RTTI range
+//      are qualified to change to their 2-byte versions.
+//      An example:
+//
+//        CompressibleRegion cr(_masm);
+//        __ andr(...);      // this instruction could change to c.and if able to
+//
+//   4. Using -XX:PrintAssemblyOptions=no-aliases could print RVC instructions instead of
+//      normal ones.
+//
+
+private:
+  bool _in_compressible_region;
+public:
+  bool in_compressible_region() const { return _in_compressible_region; }
+  void set_in_compressible_region(bool b) { _in_compressible_region = b; }
+public:
+
+// RVC: a compressible region
+  class CompressibleRegion : public StackObj {
+  protected:
+    Assembler *_masm;
+    bool _saved_in_compressible_region;
+  public:
+    CompressibleRegion(Assembler *_masm)
+    : _masm(_masm)
+    , _saved_in_compressible_region(_masm->in_compressible_region()) {
+      _masm->set_in_compressible_region(true);
+    }
+    ~CompressibleRegion() {
+      _masm->set_in_compressible_region(_saved_in_compressible_region);
+    }
+  };
+
+  // RVC: extract a 16-bit instruction.
+  static inline uint16_t c_extract(uint16_t val, unsigned msb, unsigned lsb) {
```

Comment:
Looks like c_extract and c_sextract are not used?

Reply:
Thanks for pointing out this - this seems to be inherited from the original patch (the version that we decode instructions in `emit()`) and they seem to be not used anymore and I have removed them.